c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,318 @@ WebSocketPipe newWebSocketPipe() {
+
+    auto params = splitPartsWithCondition(offer, ';');
+
+    for (const auto& p : params) {
+      if (validateTokenTail(p) != nullptr) {
+        // One of the tokens is followed by junk characters, the server will decline this
+        // offer anyways, so there's no reason to retain anyways.
+        parameters = nullptr;
+        break;
+      }
+      auto& config = KJ_REQUIRE_NONNULL(parameters);
+      if (p.startsWith("client_max_window_bits")) {
+        KJ_SWITCH_ONEOF(extractMaxBits(p)) {
+          KJ_CASE_ONEOF(maybeWindow, kj::Maybe<size_t>) {
+            KJ_IF_MAYBE(window, maybeWindow) {
+              config.outbound_max_window_bits = *window;
+            }
+          }
+          KJ_CASE_ONEOF(e, kj::Exception) {
+            // Received invalid ABNF -- server will decline so we won't bother retaining.
+            parameters = nullptr;
+          }
+        }
+        if (parameters == nullptr) {
+          break;
+        }
+      } else if (p.startsWith("client_no_context_takeover")) {
+        config.outbound_no_context_takeover = true;
+      }
+    }
+    return kj::mv(parameters);
+  }
+
+  kj::Maybe<CompressionParameters> validateCompressionParameters(
+      kj::ArrayPtr<kj::String> parameters, bool isRequest) {
+    // Given an array of parameters, populate `CompressionParameters` if the configuration is valid.
+    // `isRequest` tells us if we're parsing an offer (Request) or an agreement (Response).
+    if (parameters.size() > 5) {
+      // `permessage-deflate` has 4 optional parameters. If we've seen all 4 parameters, then any
+      // additional parameters are either repeats, or invalid, so there's no need to process them.
+      // `parameters` starts with `permessage-deflate`, so we should have 5 elements at most.
+      return nullptr;
+    }
+
+    kj::Maybe<CompressionParameters> compressionParameters = CompressionParameters {};
+    auto& config = KJ_REQUIRE_NONNULL(compressionParameters);
+
+    for (auto& i : kj::indices(parameters)) {
+      if (validateTokenTail(parameters[i]) != nullptr) {
+        // `parameters[i]` might starts with a valid extension/parameter token, but it is followed
+        // by characters that invalidate the token.
+        compressionParameters = nullptr;
+        break;
+      }
+
+      if (i == 0) {
+        // We skip the first element, `permessage-deflate`.
+        continue;
+      }
+      if (parameters[i].startsWith("server_max_window_bits")) {
+        kj::Maybe<size_t>& serverBits =
+            isRequest ? config.outbound_max_window_bits : config.inbound_max_window_bits;
+        if (serverBits != nullptr) {
+          // We've seen this parameter already, decline.
+          compressionParameters = nullptr;
+          break;
+        }
+
+        KJ_SWITCH_ONEOF(extractMaxBits(parameters[i])) {
+          KJ_CASE_ONEOF(maybeWindow, kj::Maybe<size_t>) {
+            KJ_IF_MAYBE(window, maybeWindow) {
+              serverBits = *window;
```

Comment:
Previously, if the server saw `8`, it would reply with `9`. This is wrong, we should reply with `8` and then pass `9` to zlib: https://github.com/capnproto/capnproto/pull/1480/commits/417cbbb073af297d355372b1ee514cb43c1f6adf